sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -753,6 +753,9 @@ object LimitPushDown extends Rule[LogicalPlan] {
     // Merge offset value and limit value into LocalLimit and pushes down LocalLimit through Offset.
     case LocalLimit(le, Offset(oe, grandChild)) =>
       Offset(oe, LocalLimit(Add(le, oe), grandChild))
+    // Push down local limit 1 if join type is LeftSemiOrAnti and join condition is empty.
+    case j @ Join(_, right, LeftSemiOrAnti(_), None, _) if !right.maxRows.exists(_ <= 1) =>
```

Comment:
In this case, 
when right.maxRows > 0
- for SEMI join, can we eliminate the right side and the join entirely? 
- for ANTI, can we rewrite it to an empty Relation?

when when right.maxRows = 0
- for SEMI, can we rewrite it to an empty Relation?
- for ANTI join, can we eliminate the right side and the join entirely?

Reply:
when right.maxRows > 0. It can not guarantee the result must not be empty.
when right.maxRows = 0. We can enhance `PropagateEmptyRelation` to support this case.

Reply:
```scala
    import org.apache.spark.sql.catalyst.plans.logical.Join
    val df = spark.sql("select * from range(5) t1 left semi join (select * from range(10) where id < 0) t2")
    println(df.queryExecution.optimizedPlan.asInstanceOf[Join].right.maxRows)
```
The `right.maxRows` is `Some(10)` but actually rows is 0.